actors/builtin/miner/miner_state.go

```diff
@@ -875,68 +899,64 @@ func (st *State) UnlockUnvestedFunds(store adt.Store, currEpoch abi.ChainEpoch,
 // Unlocks all vesting funds that have vested before the provided epoch.
 // Returns the amount unlocked.
 func (st *State) UnlockVestedFunds(store adt.Store, currEpoch abi.ChainEpoch) (abi.TokenAmount, error) {
-	vestingFunds, err := adt.AsArray(store, st.VestingFunds)
+	vestingFunds, err := st.LoadVestingFunds(store)
+	if err != nil {
+		return big.Zero(), xerrors.Errorf("failed to load vesting funds: %w", err)
+	}
+
+	amountUnlocked := st.unlockVestedFunds(vestingFunds, currEpoch)
+
+	err = st.SaveVestingFunds(store, vestingFunds)
 	if err != nil {
-		return big.Zero(), err
+		return big.Zero(), xerrors.Errorf("failed to save vesing funds: %w", err)
 	}
 
+	return amountUnlocked, nil
+}
+
+func (st *State) unlockVestedFunds(vestingFunds *VestingFunds, currEpoch abi.ChainEpoch) abi.TokenAmount {
 	amountUnlocked := abi.NewTokenAmount(0)
-	lockedEntry := abi.NewTokenAmount(0)
-	var toDelete []uint64
-	var finished = fmt.Errorf("finished")
-
-	// Iterate vestingFunds  in order of release.
-	err = vestingFunds.ForEach(&lockedEntry, func(k int64) error {
-		if k < int64(currEpoch) {
-			amountUnlocked = big.Add(amountUnlocked, lockedEntry)
-			toDelete = append(toDelete, uint64(k))
-		} else {
-			return finished // stop iterating
+
+	lastIndexToRemove := -1
+	for i, vf := range vestingFunds.Funds {
+		if vf.Epoch >= currEpoch {
+			break
 		}
-		return nil
-	})
 
-	if err != nil && err != finished {
-		return big.Zero(), err
+		amountUnlocked = big.Add(amountUnlocked, vf.Amount)
+		lastIndexToRemove = i
 	}
 
-	err = vestingFunds.BatchDelete(toDelete)
-	if err != nil {
-		return big.Zero(), errors.Wrapf(err, "failed to delete locked fund during vest: %v", err)
+	// remove all entries upto and including lastIndexToRemove
+	if lastIndexToRemove != -1 {
+		vestingFunds.Funds = vestingFunds.Funds[lastIndexToRemove+1:]
 	}
 
 	st.LockedFunds = big.Sub(st.LockedFunds, amountUnlocked)
```

Comment:
This still makes me a bit uncomfortable. Can we just:

1. Add an `unlockVestedFunds` method on `vestingFunds`.
2. Call that wherever this is called, making sure to update `st.LockedFunds` on the state object at the same time.
3. Get rid of this method.

Reply:
If we do that we should move all the lock/unlock manipulations to the VestingFunds object. Those methods can't update the LockedFunds total, but by making these state methods much smaller, it would be clearer that we are getting that right. I'm ok either way.

Reply:
The whole rearrangement is done.